modules/puppetca/puppetca_main.rb

```diff
@@ -16,51 +18,48 @@ def clean certname
       puppetca("clean", certname)
     end
 
-    #remove certname from autosign if exists
-    def disable certname
-      raise "No such file #{autosign_file}" unless File.exist?(autosign_file)
-
-      found = false
-      entries = File.readlines(autosign_file).collect do |l|
-        if l.chomp != certname
-          l
-        else
-          found = true
-          nil
-        end
-      end.uniq.compact
-      if found
-        open(autosign_file, File::TRUNC|File::RDWR) do |autosign|
-          autosign.write entries.join
-        end
-        logger.debug "Removed #{certname} from autosign"
-      else
-        logger.debug "Attempt to remove nonexistent client autosign for #{certname}"
-        raise NotPresent, "Attempt to remove nonexistent client autosign for #{certname}"
+    # decide if csr should be autosigned
+    # parameter is csr to use
+    def autosign csr
+      if csr.nil?
+        logger.debug "Request did not include a CSR"
+        return false
+      end
+      return true if signall
+      begin
+        req = Proxy::PuppetCa::CertificateRequest.new csr
+      rescue OpenSSL::X509::RequestError
+        logger.debug "Invalid CSR"
+        return false
       end
+      token = req.challenge_password
+      if token.nil?
+        logger.debug "CSR did not include a token"
+        return false
+      end
+      foreman_callback token
     end
 
-    # add certname to puppet autosign file
-    # parameter is certname to use
-    def autosign certname
-      FileUtils.touch(autosign_file) unless File.exist?(autosign_file)
-
-      open(autosign_file, File::RDWR) do |autosign|
-        # Check that we don't have that host already
-        found = autosign.readlines.find { |line| line.chomp == certname }
-        autosign.puts certname unless found
-      end
-      logger.debug "Added #{certname} to autosign"
+    def signall
+      Proxy::PuppetCa::Plugin.settings.signall
     end
 
-    # list of hosts which are now allowed to be installed via autosign
-    def autosign_list
-      return [] unless File.exist?(autosign_file)
-      File.read(autosign_file).split("\n").reject do |v|
-        v =~ /^\s*#.*|^$/ ## Remove comments and empty lines
-      end.map do |v|
-        v.chomp ## Strip trailing spaces
+    def foreman_callback token
+      logger.debug "Calling foreman with token '#{token}'"
+      fr = Proxy::HttpRequest::ForemanRequest.new
+      req = fr.request_factory.create_delete("api/puppetca_token", :id => token)
```

Comment:
I would suggested splitting the check and cleanup into two separate actions: not only removal of data on check not something that is expected, but it might make sense to keep the token until the host has been successfully created.

Reply:
Removing the token is part of the design. For security reasons, it should just be able to use the token once.

Reply:
In what scenario do you think would it make sense to still have the token after the CSR was verified with it? If anything goes wrong during the verification, the token will not be deleted, so we just deleted it when the host has his cert signed, when we don't need the token anymore.

Reply:
It felt to me that this is a call with a (big) side-effect, which also broke the pattern of how we use "delete" method in smart-proxy.

If you renamed ```foreman_callback``` to something like ```delete_token```, this would resolve the whole side-effect issue.